server/market/market.go

```diff
@@ -3,30 +3,435 @@
-// - Putting incoming orders into the current epoch queue, which must implement
-//   matcher.Booker so that the order matching engine can work with it.
-// - Possess an order book manager, which must also implement matcher.Booker.
+//   utxos, sufficient market buy buffer, etc.).
+// - Putting incoming orders into the current epoch queue.
+// - Maintain an order book, which must also implement matcher.Booker.
 // - Initiate order matching via matcher.Match(book, currentQueue)
 // - During and/or after matching:
 //     * update the book (remove orders, add new standing orders, etc.)
 //     * retire/archive the epoch queue
 //     * publish the matches (and order book changes?)
 //     * initiate swaps for each match (possibly groups of related matches)
-// - Continually update the order book based on data from the swap executors
-//   (e.g. failed swaps, partial fills?, etc.), communications hub (i.e. dropped
-//   clients), and other sources.
+// - Cycle the epochs.
 // - Recording all events with the archivist
+type Market struct {
+	// ctx is provided to the Market as the first argument of the constructor
+	// and subsequently provided to internal (unexported) methods as a field of
+	// Market for the sole purpose of providing a cancellation mechanism.
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
 
-// The Market manager should not be overly involved with details of accounts and
-// authentication. Via the account package it should request account status with
-// new orders, verification of order signatures. The Market should also perform
-// various account package callbacks such as order status updates so that the
-// account package code can keep various data up-to-date, including order
-// status, history, cancellation statistics, etc.
-type Market struct{} // TODO
+	// Communications.
+	orderRouter chan *orderUpdateSignal  // incoming orders
+	orderFeeds  []chan *bookUpdateSignal // outgoing notifications
+
+	// Order handling: book, active epoch queues, order matcher, and swapper.
+	running       chan struct{}
+	marketInfo    *types.MarketInfo
+	lotSize       uint64
+	bookMtx       sync.Mutex
+	book          *book.Book
+	epochDuration int64
+	epochQueue    *EpochQueue
+	matcher       *matcher.Matcher
+	swapper       *swap.Swapper
+	auth          AuthManager
+
+	// The asset backends. Presently UNUSED.
+	baseAsset  *asset.Asset
+	quoteAsset *asset.Asset
+
+	// Persistent data storage.
+	storage db.DEXArchivist
+}
+
+// NewBookUpdateChan provides a new order book update channel.
+func (m *Market) NewBookUpdateChan() <-chan *bookUpdateSignal {
+	bookUpdates := make(chan *bookUpdateSignal, 1)
+	m.orderFeeds = append(m.orderFeeds, bookUpdates)
+	return bookUpdates
+}
+
+type orderUpdateSignal struct {
+	rec     *orderRecord
+	errChan chan error // should be buffered
+
+	// sTime should not be set by the submitter
+	sTime *time.Time // initial time the order was received by the main runEpochs loop
```

Comment:
This is presently used to assist epoch order processing.  Another place for this may be desired.